File: legacy/server/routes/trading.go
package routes

import (
	"net/http"
	"saturday-autotrade/models"
	"saturday-autotrade/services"
	"strconv"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
)

func SetupTradingRoutes(router *gin.Engine) {
	api := router.Group("/api/trading")

	tradingService := services.NewTradingService()
	connectionService := services.NewConnectionService()

	// Generate trading signal endpoint (already exists in main.go, will be moved here)
	api.POST("/generate-signal", func(c *gin.Context) {
		var req models.GenerateSignalRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}

		if req.Model == "" {
			req.Model = "gpt-3.5-turbo"
		} else {
			validModels := map[string]bool{
				"gpt-3.5-turbo": true,
				"gpt-4":         true,
				"gpt-4-turbo":   true,
				"gpt-4o":        true,
				"gpt-4o-mini":   true,
				"gpt-4.1":       true}
			if !validModels[req.Model] {
				req.Model = "gpt-3.5-turbo"
			}
		}

		// Default to 1h if not provided
		selectedTimeframes := req.Timeframes
		if len(selectedTimeframes) == 0 {
			selectedTimeframes = []string{"1h"}
		}

		signal, err := tradingService.GenerateTradingSignalFromAI(req.Symbol, req.Model, selectedTimeframes)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		// Save the signal to database
		if err := tradingService.SaveTradingSignal(signal); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to save trading signal"})
			return
		}

		response := models.GenerateSignalResponse{
			Signal: signal.ToResponse(),
		}

		c.JSON(http.StatusOK, response)
	})

	// Execute trading signal endpoint
	api.POST("/execute", func(c *gin.Context) {
		var req models.ExecuteTradeRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}

		// Get the signal from database to ensure it exists and get full data
		signal, err := tradingService.GetTradingSignalByID(req.Signal.ID)
		if err != nil {
			c.JSON(http.StatusNotFound, gin.H{"error": "Trading signal not found"})
			return
		}

		// Execute the trade
		result, err := tradingService.ExecuteTrade(signal, req.IsTestnet)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		if !result.Success {
			c.JSON(http.StatusBadRequest, gin.H{
				"error":   result.Message,
				"success": false,
			})
			return
		}

		c.JSON(http.StatusOK, result)
	})

	// Execute manual JSON signal endpoint
	api.POST("/execute-manual", func(c *gin.Context) {
		defer func() {
			if r := recover(); r != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
			}
		}()

		var req models.ExecuteManualSignalRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}

		// Validate that signal JSON is not empty
		if strings.TrimSpace(req.SignalJson) == "" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Signal JSON is required"})
			return
		}

		// Execute the manual signal
		result, err := tradingService.ExecuteManualSignal(req.SignalJson, req.IsTestnet)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		if !result.Success {
			c.JSON(http.StatusBadRequest, gin.H{
				"error":   result.Message,
				"success": false,
			})
			return
		}

		c.JSON(http.StatusOK, result)
	})

	// Fetch trading signals endpoint
	api.GET("/signals", func(c *gin.Context) {
		limitStr := c.DefaultQuery("limit", "50")
		limit, err := strconv.Atoi(limitStr)
		if err != nil {
			limit = 50
		}

		signals, err := tradingService.GetTradingSignals(limit)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		// Convert to response format
		var signalResponses []models.TradingSignalResponse
		for _, signal := range signals {
			signalResponses = append(signalResponses, signal.ToResponse())
		}

		c.JSON(http.StatusOK, gin.H{"signals": signalResponses})
	})

	// Get single trading signal by ID
	api.GET("/signals/:id", func(c *gin.Context) {
		id := c.Param("id")

		signal, err := tradingService.GetTradingSignalByID(id)
		if err != nil {
			c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
			return
		}

		c.JSON(http.StatusOK, gin.H{"signal": signal.ToResponse()})
	})

	// Get Binance price endpoint
	api.GET("/binance-price/:symbol", func(c *gin.Context) {
		defer func() {
			if r := recover(); r != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
			}
		}()

		symbol := c.Param("symbol")

		price, err := tradingService.GetBinancePrice(symbol)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		c.JSON(http.StatusOK, price)
	})

	// Create new trading position
	api.POST("/positions", func(c *gin.Context) {
		defer func() {
			if r := recover(); r != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
			}
		}()

		var req models.CreatePositionRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}

		// Validate required fields
		if req.Symbol == "" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Symbol is required"})
			return
		}

		if req.Direction != "LONG" && req.Direction != "SHORT" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Direction must be LONG or SHORT"})
			return
		}

		if req.Size <= 0 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Size must be greater than 0"})
			return
		}

		if req.EntryPrice <= 0 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Entry price must be greater than 0"})
			return
		}

		if req.Leverage < 1 || req.Leverage > 125 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Leverage must be between 1 and 125"})
			return
		}

		// Create the position
		position, err := tradingService.CreatePosition(&req)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		response := models.CreatePositionResponse{
			Success:  true,
			Position: position.ToResponse(),
			Message:  "Successfully created position",
		}

		c.JSON(http.StatusCreated, response)
	})

	// Get all trading positions
	api.GET("/positions", func(c *gin.Context) {
		defer func() {
			if r := recover(); r != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
			}
		}()

		positions, err := tradingService.GetPositions()
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		// Convert to response format and ensure we always have a slice, never nil
		positionResponses := make([]models.PositionResponse, 0, len(positions))
		for _, position := range positions {
			positionResponses = append(positionResponses, position.ToResponse())
		}

		response := gin.H{"positions": positionResponses}

		c.JSON(http.StatusOK, response)
	})

	// Close trading position
	api.POST("/positions/:id/close", func(c *gin.Context) {
		defer func() {
			if r := recover(); r != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
			}
		}()

		id := c.Param("id")

		var req models.ClosePositionRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}

		// Validate close price
		if req.ClosePrice <= 0 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Close price must be greater than 0"})
			return
		}

		// Close the position (service currently only supports id)
		position, err := tradingService.ClosePosition(id)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		response := models.ClosePositionResponse{
			Success:  true,
			Position: position.ToResponse(),
			Message:  "Successfully closed position",
		}

		c.JSON(http.StatusOK, response)
	})

	// Get transaction history
	api.GET("/transactions", func(c *gin.Context) {
		defer func() {
			if r := recover(); r != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
			}
		}()

		limitStr := c.DefaultQuery("limit", "50")
		limit, err := strconv.Atoi(limitStr)
		if err != nil {
			limit = 50
		}

		transactions, err := tradingService.GetTransactions(limit)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		// Convert to response format and ensure we always have a slice, never nil
		transactionResponses := make([]models.TransactionResponse, 0, len(transactions))
		for _, transaction := range transactions {
			transactionResponses = append(transactionResponses, transaction.ToResponse())
		}

		c.JSON(http.StatusOK, gin.H{"transactions": transactionResponses})
	})

	// Get performance metrics
	api.GET("/performance", func(c *gin.Context) {
		defer func() {
			if r := recover(); r != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
			}
		}()

		metrics, err := tradingService.GetPerformanceMetrics()
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		c.JSON(http.StatusOK, gin.H{"metrics": metrics})
	})

	// Get connection status endpoint
	api.GET("/status", func(c *gin.Context) {
		defer func() {
			if r := recover(); r != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
			}
		}()

		status := connectionService.GetConnectionStatus()

		c.JSON(http.StatusOK, gin.H{"status": status})
	})

	// Get real-time price data endpoint (new format)
	api.GET("/prices/:symbol", func(c *gin.Context) {
		defer func() {
			if r := recover(); r != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
			}
		}()

		symbol := c.Param("symbol")

		// Validate symbol format - should be uppercase and likely end with USDT
		if len(symbol) < 3 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid symbol format"})
			return
		}

		// Convert symbol to standard format (e.g., BTC -> BTCUSDT)
		standardSymbol := symbol
		if symbol == "BTC" {
			standardSymbol = "BTCUSDT"
		} else if symbol == "ETH" {
			standardSymbol = "ETHUSDT"
		} else if symbol == "XRP" {
			standardSymbol = "XRPUSDT"
		} else if symbol == "DOGE" {
			standardSymbol = "DOGEUSDT"
		} else if !strings.HasSuffix(strings.ToUpper(symbol), "USDT") {
			standardSymbol = strings.ToUpper(symbol) + "USDT"
		}

		binanceService := services.NewBinanceService()
		priceData, err := binanceService.GetPrice(standardSymbol)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch price data for symbol"})
			return
		}

		// Create the response with all required fields
		response := gin.H{
			"symbol":           symbol,
			"currentPrice":     priceData.Price,
			"timestamp":        time.Now().UTC().Format(time.RFC3339),
			"volume24h":        priceData.Volume,
			"percentChange24h": priceData.Change24h,
		}

		c.JSON(http.StatusOK, response)
	})

	// Get real USDT balance endpoint
	api.GET("/balance", func(c *gin.Context) {
		futuresService := services.NewBinanceFuturesService(false) // false = mainnet
		accountInfo, err := futuresService.GetAccountInfo()
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
		usdtBalance := 0.0
		for _, asset := range accountInfo.Assets {
			if asset.Asset == "USDT" {
				usdtBalance, _ = strconv.ParseFloat(asset.WalletBalance, 64)
				break
			}
		}
		c.JSON(http.StatusOK, gin.H{"usdtBalance": usdtBalance})
	})

	// Chart data prompt endpoint
	api.POST("/chart-data-prompt", func(c *gin.Context) {
		var req struct {
			Symbol     string   `json:"symbol"`
			Timeframes []string `json:"timeframes"`
		}
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request"})
			return
		}
		prompt, err := tradingService.BuildChartDataPrompt(req.Symbol, req.Timeframes)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, gin.H{"prompt": prompt})
	})
}

File: legacy/client/src/api/trading.ts
import api from './api';
import { TradingSignal, Position, Transaction, PerformanceMetrics, TradingConfig, ConnectionStatus } from '../types/trading';

console.log('trading.ts: Imported api:', typeof api, !!api);

// Description: Get current price from Binance
// Endpoint: GET /api/trading/binance-price/:symbol
// Request: {}
// Response: { price: number, change24h: number, volume: number }
export const getBinancePrice = async (symbol: string) => {
  console.log('getBinancePrice: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in getBinancePrice');
  }

  try {
    const response = await api.get(`/api/trading/binance-price/${symbol}`);
    return response.data;
  } catch (error: any) {
    console.error(error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Generate AI trading signal
// Endpoint: POST /api/trading/generate-signal
// Request: { symbol: string, model: string, timeframes: string[] }
// Response: { signal: TradingSignal }
export const generateTradingSignal = async (data: { symbol: string; model: string; timeframes?: string[] }) => {
  console.log('generateTradingSignal: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in generateTradingSignal');
  }

  try {
    const response = await api.post('/api/trading/generate-signal', data);
    return response.data;
  } catch (error: any) {
    console.error(error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Get trading signals
// Endpoint: GET /api/trading/signals
// Request: { limit?: number }
// Response: { signals: TradingSignal[] }
export const getTradingSignals = async (limit = 50) => {
  console.log('getTradingSignals: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in getTradingSignals');
  }

  try {
    console.log('getTradingSignals: Making API call to /api/trading/signals');
    const response = await api.get(`/api/trading/signals?limit=${limit}`);
    console.log('getTradingSignals: Received response:', response.data);
    return response.data;
  } catch (error: any) {
    console.error('getTradingSignals: Error occurred:', error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Execute trading signal
// Endpoint: POST /api/trading/execute
// Request: { signal: TradingSignal, isTestnet: boolean }
// Response: { success: boolean, transactionId: string, message?: string }
export const executeTrade = async (data: { signal: TradingSignal; isTestnet: boolean }) => {
  console.log('executeTrade: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in executeTrade');
  }

  try {
    const response = await api.post('/api/trading/execute', data);
    return response.data;
  } catch (error: any) {
    console.error(error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Get current open positions
// Endpoint: GET /api/trading/positions
// Request: {}
// Response: { positions: Position[] }
export const getPositions = async () => {
  console.log('getPositions: Starting function...');
  console.log('getPositions: API instance check:', typeof api, !!api);

  if (!api) {
    console.error('getPositions: API instance is undefined/null!');
    throw new Error('API instance is not available in getPositions');
  }

  try {
    console.log('getPositions: Making API call to /api/trading/positions');
    const response = await api.get('/api/trading/positions');
    console.log('getPositions: Received response status:', response.status);
    console.log('getPositions: Received response data:', response.data);
    return response.data;
  } catch (error: any) {
    console.error('getPositions: Error occurred:', error);
    console.error('getPositions: Error message:', error.message);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Create a new trading position
// Endpoint: POST /api/trading/positions
// Request: { symbol: string, direction: 'LONG' | 'SHORT', size: number, entryPrice: number, leverage: number, isTestnet: boolean, stopLoss?: number, takeProfit?: number }
// Response: { success: boolean, position: Position, message: string }
export const createPosition = async (data: {
  symbol: string;
  direction: 'LONG' | 'SHORT';
  size: number;
  entryPrice: number;
  leverage: number;
  isTestnet: boolean;
  stopLoss?: number;
  takeProfit?: number;
}) => {
  console.log('createPosition: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in createPosition');
  }

  try {
    const response = await api.post('/api/trading/positions', data);
    return response.data;
  } catch (error: any) {
    console.error('Error creating position:', error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Close a trading position
// Endpoint: POST /api/trading/positions/:id/close
// Request: { closePrice: number }
// Response: { success: boolean, position: Position, message: string }
export const closePosition = async (positionId: string, data: { closePrice: number }) => {
  console.log('closePosition: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in closePosition');
  }

  try {
    console.log('closePosition: Making API call to close position:', positionId);
    const response = await api.post(`/api/trading/positions/${positionId}/close`, data);
    console.log('closePosition: Received response:', response.data);
    return response.data;
  } catch (error: any) {
    console.error('closePosition: Error occurred:', error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Get transaction history
// Endpoint: GET /api/trading/transactions
// Request: { limit?: number }
// Response: { transactions: Transaction[] }
export const getTransactions = async (limit = 50) => {
  console.log('getTransactions: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in getTransactions');
  }

  try {
    console.log('getTransactions: Making API call to /api/trading/transactions');
    const response = await api.get(`/api/trading/transactions?limit=${limit}`);
    console.log('getTransactions: Received response:', response.data);
    return response.data;
  } catch (error: any) {
    console.error('getTransactions: Error occurred:', error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Get performance metrics
// Endpoint: GET /api/trading/performance
// Request: {}
// Response: { metrics: PerformanceMetrics }
export const getPerformanceMetrics = async () => {
  console.log('getPerformanceMetrics: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in getPerformanceMetrics');
  }

  try {
    console.log('getPerformanceMetrics: Making API call to /api/trading/performance');
    const response = await api.get('/api/trading/performance');
    console.log('getPerformanceMetrics: Received response:', response.data);
    return response.data;
  } catch (error: any) {
    console.error('getPerformanceMetrics: Error occurred:', error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Get connection status
// Endpoint: GET /api/trading/status
// Request: {}
// Response: { status: ConnectionStatus }
export const getConnectionStatus = async () => {
  console.log('getConnectionStatus: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in getConnectionStatus');
  }

  try {
    console.log('getConnectionStatus: Making API call to /api/trading/status');
    const response = await api.get('/api/trading/status');
    console.log('getConnectionStatus: Received response:', response.data);
    return response.data;
  } catch (error: any) {
    console.error('getConnectionStatus: Error occurred:', error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Get current price for symbol
// Endpoint: GET /api/trading/price/:symbol
// Request: {}
// Response: { price: number, change24h: number }
export const getCurrentPrice = async (symbol: string) => {
  console.log('getCurrentPrice: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in getCurrentPrice');
  }

  try {
    const response = await api.get(`/api/trading/binance-price/${symbol}`);
    return {
      price: response.data.price,
      change24h: response.data.change24h
    };
  } catch (error: any) {
    console.error('Error fetching current price:', error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Execute manual JSON signal
// Endpoint: POST /api/trading/execute-manual
// Request: { signalJson: string, isTestnet: boolean }
// Response: { success: boolean, signal: TradingSignal, transactionId: string, message?: string }
export const executeManualSignal = async (data: { signalJson: string; isTestnet: boolean }) => {
  console.log('executeManualSignal: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in executeManualSignal');
  }

  try {
    console.log('executeManualSignal: Making API call to /api/trading/execute-manual');
    const response = await api.post('/api/trading/execute-manual', data);
    console.log('executeManualSignal: Received response:', response.data);
    return response.data;
  } catch (error: any) {
    console.error('executeManualSignal: Error occurred:', error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Get real-time price data for cryptocurrency
// Endpoint: GET /api/trading/prices/:symbol
// Request: {}
// Response: { symbol: string, currentPrice: number, timestamp: string, volume24h: number, percentChange24h: number }
export const getRealTimePriceData = async (symbol: string) => {
  console.log('getRealTimePriceData: API instance check:', typeof api, !!api);
  if (!api) {
    throw new Error('API instance is not available in getRealTimePriceData');
  }

  try {
    console.log('getRealTimePriceData: Making API call to /api/trading/prices/' + symbol);
    const response = await api.get(`/api/trading/prices/${symbol}`);
    console.log('getRealTimePriceData: Received response:', response.data);
    return response.data;
  } catch (error: any) {
    console.error('getRealTimePriceData: Error occurred:', error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};

// Description: Get real USDT balance
// Endpoint: GET /api/trading/balance
// Response: { usdtBalance: number }
export const getUsdtBalance = async () => {
  if (!api) throw new Error('API instance is not available in getUsdtBalance');
  try {
    const response = await api.get('/api/trading/balance');
    return response.data;
  } catch (error: any) {
    console.error('getUsdtBalance: Error occurred:', error);
    throw new Error(error?.response?.data?.error || error.message);
  }
};